### DISCUSSION TITLE
Named destinations (PDF's anchors)

### DISCUSSION THREAD

**Discussion by nv2p** [vedang/pdf-tools#156] [Ideas]
Hello everyone and thanks for this amazing software!

I like precise and sustainable references and got aware of named destinations in PDFs. Adobe Acrobat Pro or PDF Studio can create them and browsers recognize them. Okular or Evince, on the other hand, have a bookmarking system that don't create named destinations (even their bookmarks are not recognized elsewhere).

I don't know how to implement this and if there're standards in this field, but I saw that poppler is aware of such named destinations (`pdfinfo -dests input.pdf` show them). I also noticed that it's possible, if we know the id (e.g. through the table of contents), to bring a browser directly to a chapter by adding it's id after a #. In a file of mine, e.g., one chapter has the id #[{%22num%22%3A25%2C%22gen%22%3A0}%2C{%22name%22%3A%22XYZ%22}%2C59%2C714%2C0].

I'm aware of precise notes in org-noter, but as I understand, it's an emacs (org-mode) specific way of referring and it depends on pages. Bookmarks are more reliables, but inside emacs and their purpose is not to be used in documents. I'm no programmer, but it seems to me that using marked destination, if possible, would be very nice to have annotations/markups, which we can refer to precisely (e.g. see my comment #xzy) _and_ share. If org files are used to annotate, they could be converted, with the exact position. Also just as personnal convenience, I would enjoy to can just click on my reference and be brought to the exact position, regardless of if I use pdf-tools or not (the same way that I like to can read my org files outside emacs, when needed).

I didn't find anything here with the keyword "anchor" or "named destination" and "bookmark" always refer to the emacs's ones, so I would like to know if someone thought/worked on this or how you deal with precise references. I can imagine it isn't trivial, maybe not possible at the moment, so I chose not to open an issue, but a discussion, just to share. My ideal would be a pdf-tools where I can highlight a text, hit a shortcut and the function 1) highlight the text, 2) give an named destination to it 3) put this id in the annotation field of the highlight, ready for me to add my comment [4) C-u to save the reference with the id? :)]. The third point is because if the page has many comments, it's better to find it. If then pdf-tools would put an arrow on the right place, like it does with chapters, it would be the cherry on the cake!

**Comment by vedang** [comment]
When you create an annotation, you generate an ID and insert it in the document at the annotation location. One way to make this easily discoverable would be to add the ability to edit annotation labels, similar to how `pdf-annot-edit-contents` allows the user to edit the contents of the annotation -- thereby allowing you to store extra details in any annotation. 

The label is an editable property of the annotation, but no functionality has been built into `pdf-tools` to allow editing it. Once such is done, then the usage would look like follows:
1. If you provide a `C-u` argument when creating the annotation, ask the user for a label name as well. 
2. Filter the list of annotations on the label using the built-in filtering functionality of `tablist` (`C-c C-a l` to list annotations, followed by `/ e` to filter them in multiple different ways)
3. Profit.

You can do what you want to today in a roundabout way, by adding some unique search term to the contents of your annotation and then following all the same steps listed above.

**Reply by nv2p** [reply]
Thanks a lot for this feedback and workaround. It's clever, but it remains an Emacs only solution and even in Emacs, it isn't so trivial to create references which bring straight to the comment. That said, if I create named destinations in another program and start the comments with this denomination, I can at least find them in Emacs this way. 

I have checked a bit more carefully the labels and there seems to be actually two kinds of labels:
1. The ones which you can jump to with the function `pdf-view-goto-label`. Those seem to usually overlap with pages and if I create _named destinations_ in another program, which are recognized by `pdfinfo -dests` (and I can jump to them with e.g. Firefox too), they don't appear when I call `pdf-view-goto-label`.
1. The ones in the annotations list. When I call `pdf-annot-list-annotations`, I have 4 columns: Pg., Type, Label and Date. When I create annotations with pdf-tools, there is nothing in the Label column, but when I create annotations with other programs, where I have a username defined, I get this username in the Label column. So those seem to correspond more to something like ownership.

I guess that your proposition would, if any, correspond more to the first kind. I would find two things important if those are implemented.
1) My main point is to have something compatible with other programs. Otherwise, unless it's quite easy to implement, I don't really find the effort worth (there is already bookmarks, org-noter, org-pdftools, etc.). I wouldn't mind much if named annotations are called labels in Emacs (though I would find it better to use the same name as others), but the highest priority in my eyes is that those "labels" should be recognized by other programs.
2) Less important, the Label column in the annotations list should contain those (first) labels and the second kind should be renamed e.g. Owner(ship), if my observation on this matter is right.
